gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnAutoScalingManager.java

```diff
@@ -0,0 +1,158 @@
+  private final String AUTO_SCALING_PARTITIONS_PER_CONTAINER = AUTO_SCALING_PREFIX + "partitionsPerContainer";
+  private final int DEFAULT_AUTO_SCALING_PARTITIONS_PER_CONTAINER = 1;
+
+  private final Config config;
+  private final HelixManager helixManager;
+  private final ScheduledExecutorService autoScalingExecutor;
+  private final YarnService yarnService;
+  private final int partitionsPerContainer;
+
+  public YarnAutoScalingManager(GobblinApplicationMaster appMaster) {
+    this.config = appMaster.getConfig();
+    this.helixManager = appMaster.getMultiManager().getJobClusterHelixManager();
+    this.yarnService = appMaster.getYarnService();
+    this.partitionsPerContainer = ConfigUtils.getInt(this.config, AUTO_SCALING_PARTITIONS_PER_CONTAINER,
+        DEFAULT_AUTO_SCALING_PARTITIONS_PER_CONTAINER);
+
+    Preconditions.checkArgument(this.partitionsPerContainer > 0,
+        AUTO_SCALING_PARTITIONS_PER_CONTAINER + " needs to be greater than 0");
+
+    this.autoScalingExecutor = Executors.newSingleThreadScheduledExecutor(
+        ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("AutoScalingExecutor")));
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    int scheduleInterval = ConfigUtils.getInt(this.config, AUTO_SCALING_POLLING_INTERVAL_SECS,
+        DEFAULT_AUTO_SCALING_POLLING_INTERVAL_SECS);
+    log.info("Starting the " + YarnAutoScalingManager.class.getSimpleName());
+    log.info("Scheduling the auto scaling task with an interval of {} seconds", scheduleInterval);
+
+    this.autoScalingExecutor.scheduleAtFixedRate(new YarnAutoScalingRunnable(new TaskDriver(this.helixManager),
+            this.yarnService, this.partitionsPerContainer), 0,
+        scheduleInterval, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    log.info("Stopping the " + YarnAutoScalingManager.class.getSimpleName());
+
+    ExecutorsUtils.shutdownExecutorService(this.autoScalingExecutor, Optional.of(log));
+  }
+
+  /**
+   * A {@link Runnable} that figures out the number of containers required for the workload
+   * and requests those containers.
+   */
+  @VisibleForTesting
+  @AllArgsConstructor
+  static class YarnAutoScalingRunnable implements Runnable {
+    private final TaskDriver taskDriver;
+    private final YarnService yarnService;
+    private final int partitionsPerContainer;
+
+    /**
+     * Iterate through the workflows configured in Helix to figure out the number of required partitions
+     * and request the {@link YarnService} to scale to the desired number of containers.
+     */
+    @Override
+    public void run() {
+      Set<String> inUseInstances = new HashSet<>();
+
+      int numPartitions = 0;
+      for (Map.Entry<String, WorkflowConfig> workFlowEntry : taskDriver.getWorkflows().entrySet()) {
+        WorkflowContext workflowContext = taskDriver.getWorkflowContext(workFlowEntry.getKey());
+
+        // Only allocate for active workflows
```

Comment:
Is the idea to allocate containers for active workflow asynchronously ? ( every 60 seconds )
shouldn't it be inline with WorkFlow execution, i.e. when workflow gets picked up for execution, it should figure out how many container it needs and allocate accordingly?
OR, is it that Gobblin will keep certain number of containers alive to reuse ?

Reply:
i see, thanks for the explanation around this.